packages/actor-query-operation-path-one-or-more/lib/ActorQueryOperationPathOneOrMore.ts

```diff
@@ -59,8 +71,41 @@ export class ActorQueryOperationPathOneOrMore extends ActorAbstractPath {
       return { type: 'bindings', bindingsStream, variables: [ objectString ]};
     }
     if (sVar && oVar) {
-      throw new Error('ZeroOrMore path expressions with 2 variables not supported yet');
-    } else if (sVar && !oVar) {
+      // Get all the results of subjects with same predicate, but once, then fill in first variable for those
+      const single = ActorAbstractPath.FACTORY.createPath(path.subject, path.predicate.path, path.object, path.graph);
+      const results = ActorQueryOperation.getSafeBindings(
+        await this.mediatorQueryOperation.mediate({ context, operation: single }),
+      );
+      const subjectString = termToString(path.subject);
+      const objectString = termToString(path.object);
+
+      const subjects: Set<string> = new Set();
+
+      const bindingsStream: MultiTransformIterator<Bindings, Bindings> = new MultiTransformIterator(
+        results.bindingsStream,
+        {
+          multiTransform: (bindings: Bindings) => {
+            if (subjects.has(termToString(bindings.get(subjectString)))) {
+              return new EmptyIterator();
+            }
+            subjects.add(termToString(bindings.get(subjectString)));
+            const val = bindings.get(subjectString);
+            return new TransformIterator<Bindings>(
+              async() => ActorQueryOperation.getSafeBindings(await this.mediatorQueryOperation.mediate({
+                context, operation: ActorAbstractPath.FACTORY.createPath(val, predicate, path.object, path.graph),
```

Comment:
Seems correct, but I'm wondering if you could use a shared subjects map for all the separate streams as is done [above](https://github.com/comunica/comunica/blob/master/packages/actor-query-operation-path-one-or-more/lib/ActorQueryOperationPathOneOrMore.ts#L35-L58). That should cut out all duplicate work that would currently be done.

Reply:
What work is duplicate in this code then? I also think it's not that easy, because above it can be passed to ALP to emit right there directly, but here it would require a change in the mediate function?

Reply:
Currently, both `a` and `b` would be in the subjects Set. This would generate 2 TransformIterators, one starting for each subject. In the first case this would call the ALP function with `a`, which would then recursively call itself again with `b`, since the ALP function starts with an empty map by default. The second stream would call the ALP function once with `b`. If they shared the same term `termHashes` either the second call to the ALP in the first stream, or the first call in the second stream wouldn't happen (depending on timing).

Reply:
Aha, allright, I see. Is it possible to call ALP with a variable as predicate?

Reply:
I'm assuming you mean subject instead of predicate? But no, you need a constant term. So your first call to generate the subjects is still necessary here. But in the streams generated in the multitransformiterator you can call ALP instead of the mediator and use the same map for al those calls.

Reply:
Yes, I figured. That's why I deleted the post :)

Reply:
I'm struggling implementing it. Directly calling ALP with the subjects is not possible, because this would generate to many solutions. So I called it with the objects, and then we transform `it` with as subject the subject of the single operation and as object the result of the ALP. But that also does not work with a centralized it, because then it's not clear which subject to add of which single operation.

Reply:
Could be that the ALP function is not the correct one to call for this and perhaps some other function would be needed (it's been a while so I don't recall exactly how ALP works). It just felt to me like this is something that should be able to be optimized since you're looking for a bunch of triples with the same predicate every time. But I'm not the one writing it right now so I could be wrong.